AC_InputManager_Heli: collective transition ramp

Ferruccio1984 opened the pull request:
Adds 0.5s ramp on the collective output when switching to (and between) manual modes.
Also involves a rework of the pre-existent stabilize mode to acro mode collective transitioning

bnsgeyer commented on libraries/AC_InputManager/AC_InputManager_Heli.cpp:
```diff
@@ -107,17 +107,34 @@ float AC_InputManager_Heli::get_pilot_desired_collective(int16_t control_in)
+        float dt = 2/(float)_loop_rate;
+        _timer -= dt;
+        _timer = constrain_float(_timer, 0.0f, 1.0f);
+        _ramp = (3*_timer*_timer)-(2*_timer*_timer*_timer);
```
please add comment here on type of function used here and why

thespova commented:
Is it too late or would it be possible to make the ramp time configurable? If its still taking collective changes at the same speed as input and just sorta sliding from what the output would be in one mode to what it would be in another mode it seems like a peram to adjust and tune this would be nice.  If the input is locked during transition than I guess this would not matter and be safer to just make it static at what ever has been tested as safe.

Im trying to learn to read more of this in code but not there yet at all.

Either way, I'm super stoked for this!

thespova commented:
I'm sure other aircraft types would benefit from this. I'm honestly shocked this isn't already part of ardupilot for all vehicles.

Here is a video from The Mach Initiative from months ago where they had a crash on a fixed wing plane they suspect is caused by the same mode transition problems that we have on heli.

TIME STAMP LINK: https://youtu.be/Fun1ahjo1bM?t=461

That said, I'm very eager to get this for heli on the test branch so I can try it out.

Ferruccio1984 commented:
@LupusTheCanine  Just to clarify, this PR aims at fixing the switch between rate controlled collective outputs and manual outputs. It is a blending between the last output of previous flight mode and the current mode output. So your statement about helis is not correct as it's not doing a rescaling, again, it's doing a blending of the outputs (even for case of switching between manual modes).

Ferruccio1984 commented:
> My somewhat general comment is that I am not a big fan of re-implementing functionality like low-pass filters in our own way in heli

Hi @MattKear I have removed that part and corrected the missing whitespaces, thanks

MattKear commented on libraries/AC_InputManager/AC_InputManager_Heli.cpp:
```diff
@@ -107,17 +107,34 @@ float AC_InputManager_Heli::get_pilot_desired_collective(int16_t control_in)
-        _stab_col_ramp += 2.0f/(float)_loop_rate;
-    } else if(!_im_flags_heli.use_stab_col && (_stab_col_ramp > 0.0f)){
-        _stab_col_ramp -= 2.0f/(float)_loop_rate;
+    if (init_transition_ramp) {
```
I think you could delete this init_transition_ramp bool.  You are already setting _ramp to 1.0.  at the same point that you set init_transisition_ramp.  See my comment later on about using is_positive(_ramp) instead.

MattKear commented on libraries/AC_InputManager/AC_InputManager_Heli.cpp:
```diff
@@ -107,17 +107,34 @@ float AC_InputManager_Heli::get_pilot_desired_collective(int16_t control_in)
 
-    // scale collective output smoothly between acro and stab col
+    // ramp function
+    if (transition) {
```
Do we need this transition bool?  Can we not use if (is_positive(_ramp)) instead?  It will cut out the extra is_zero check you have lower down.

MattKear commented on libraries/AC_InputManager/AC_InputManager_Heli.h:
```diff
@@ -46,8 +49,17 @@ class AC_InputManager_Heli : public AC_InputManager {
-    //  factor used to smoothly ramp collective from Acro value to Stab-Col value
-    float _stab_col_ramp = 0;
+    // init transition ramp
+    bool init_transition_ramp = false;
```
This is now an unused variable so can be deleted.

MattKear commented on libraries/AC_InputManager/AC_InputManager_Heli.h:
```diff
@@ -46,8 +49,17 @@ class AC_InputManager_Heli : public AC_InputManager {
+    float _old_flightmode_col_output = 0;
+
+    // ramp factor from previous mode to current mode collective output
+    float _ramp = 0;
```
I believe we try and maintain a pattern of not initialising a variable in the header file if can help it.  As you set the ramp value in the heli_init() function in heli.cpp this initialisation is not needed.
```suggestion
    float _ramp;
```

MattKear reviewed (changes requested):
This is looking good pretty good to me.  There are a couple more tidy ups that I would appreciate it if you could resolve.  Sorry I did not catch them on the last review, they just become more evident now that the code as been simplified.

Can you also squash your commits so that you only have one commit per vehicle/lib.

Have you had chance to re-test this since you have re-worked the code?

thespova commented:
Does this only slew input when going from manual to auto collective modes or all modes? IDK if this makes sense but in my head being able to use it to ramp from stabilized to acro seems like it would allow a non 3d transmitter throttle/collective curve in stabilized and then a 3d curve for acro and the pilot would have a small time phase to adapt the sticks as it switched?

Also configurable ramp time would be really good.

thespova commented:
Is this part of the "latest" on the custom build server now?  I am eager to try it out!

bnsgeyer reviewed (approved):
@MattKear Thanks for your review.  I'm happy with this PR and will merge once it passes CI.

thespova commented:
I asked earlier but didn't get an answer. I think its a pretty important question though. Does this also work between Stabilized and ACRO?

Since stabilized can have its own curve defined in arducopter as well as acro having its own expo defined in arducopter this feature would be more beneficial to transitioning between those to than any other two modes. The docs even warn about the jump to ACRO from other modes and from stab when stab has a custom curve for center hover.

https://ardupilot.org/copter/docs/traditional-helicopter-aerobatic-setup.html
